keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+
+##### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL ensure that there are json file[s] in CDI format
+for the allocated resource. These files SHALL be used by runtime to
+update runtime configuration before creating containers that use the
+resource.
+
+This operation SHALL do as little work as possible as it’s called
+after a pod is scheduled to a node. All potentially failing operations
+SHALL be done during allocation phase.
+
+This operation MUST be idempotent. If the resource corresponding to
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+On a successful call this RPC should return set of fully qualified
+CDI device names, which kubelet MUST pass to the runtime through the CRI
+protocol.
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated string cdi_device = 1;
+}
+```
+
+CRI protocol MUST be extended for this purpose, e.g. list of CDI
+device ids should be added to the CRI Device structure:
+
+```protobuf
+// Device specifies a host device to mount into a container.
+message Device {
+    ...
+    string permissions = 3;
+    // Set of fully qualified CDI device names
+    repeated string cdi_device = 4;
+}
+```
+
+###### NodePrepareResource Errors
+
+If the plugin is unable to complete the NodePrepareResource call
+successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST
+return the specified gRPC error code.  Kubelet MUST implement the
+specified error recovery behavior when it encounters the gRPC error
+code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Resource does not exist | 5 NOT_FOUND | Indicates that a resource corresponding to the specified `resource_id` does not exist. | Caller MUST verify that the `resource_id` is correct and that the resource is accessible and has not been deleted before retrying with exponential back off. |
+ 
+
+##### `NodeUnprepareResource`
+
+A Kubelet Plugin MUST implement this RPC call. This RPC is a reverse
+operation of `NodePrepareResource`. This RPC MUST undo the work by
+the corresponding `NodePrepareResource`. This RPC SHALL be called by
```

Comment:
does the kubelet need to checkpoint any state to disk to keep the SHALL constraint?

Reply:
can the kubelet ask if the resource was already prepared prior?

Reply:
Yes, it would need to track successful calls to `NodePrepareResource`. Doing the same for volumes has been problematic in the past.

Asking the plugin may help and could be added.